woocommerce/wc-admin
# Implement tabular data filtering

Fixes #1102.

This PR implements tabular data filtering for report tables. Notice there are several pending tasks/open issues detailed at the end of this post. I will open issues/create PR for them as follow-ups of this one.

### Screenshots


![search](https://user-images.githubusercontent.com/3616980/51907804-3c15be00-23c8-11e9-8f76-641fc571e0c8.gif)

### Detailed test instructions:
- Go to the _Products_ report.
- Enter some product names in the table search bar and verify filtering works as expected.
- Try deleting them and adding new ones.
- Try adding _Single Product_, _Comparison_, etc. filters and verify table searches are reset.
- While having _Single Product_, _Comparison_, etc. filters on, start searching with the table search bar and verify filters are reset.
- Trying searching partial strings, for example `sun` and verify `Sunglasses` appear.
- Try adding search strings directly to the URL after the `search=` parameter, separated with a comma, and verify the table is updated with the correct results.
- Try refreshing the page and verify searches are preserved.
- Repeat the process above with all other reports that have search bars in the table: Categories, Coupons, Taxes and Customers. Notice some are not working yet (see section below).

### What is not working? & Pending tasks
In order to avoid this PR growing too much and to get some early feedback before continuing, I will leave some work for follow-ups:
- The _Clear all_ from designs is implemented in: #1410.
- `variations` autocompleter is not working (if I understood it correctly, it will be fixed with #1391). That makes searching in the variations table not to work either.
- The Rest API call used to populate the autocompleter is not reused later to get the IDs to populate the table, causing two different API requests to be called when we really only need one (#1453).
- The app was crashing when a search was returning 0 results. For now, I made the app to ignore the search in those cases and added a `@TODO` [comment](https://github.com/woocommerce/wc-admin/pull/1381/files#diff-2459417b12e8fe2d482313c6d8f9819aR151) (#1451).
- Current `products` wc endpoint doesn't allow searching by `sku`.
- Current `customer` wc endpoint doesn't allow searching by `name + ' ' + last_name` (#1455).
- Current `customer`  wc-admin endpoint doesn't allow limiting the results by customer ID (#1457).
- Current `taxes`  wc-admin endpoint doesn't allow limiting the results by tax ID (#1456).
- I created a generic `getItems( type )` method and migrated `getCategories()` to use it. It would make sense to migrate `getOrders()` to `getItems( 'orders' )` as well in order to DRY (#1438).
- Handle the case when there is more than one page of results (#1454).
